_source/_posts/2023-03-01-jakarta-ee-oidc.md

```diff
@@ -0,0 +1,497 @@
+
+## Jakarta EE example project overview
+
+I won't reproduce the entire `pom.xml` file here, but I want to point out a few things.
+
+First, take a look at the dependencies. The only dependency required for Jakarta EE is the first one (`jakarta.jakartaee-api`). 
+
+```xml
+<dependencies>
+    <dependency>
+        <groupId>jakarta.platform</groupId>
+        <artifactId>jakarta.jakartaee-api</artifactId>
+        <version>${jakartaee-api.version}</version>
+        <scope>provided</scope>
+    </dependency>
+</dependencies>
+```
+
+The WildFly plugin is included and configured in the block below.
+
+```xml
+<plugin>
+    <groupId>org.wildfly.plugins</groupId>
+    <artifactId>wildfly-maven-plugin</artifactId>
+    <version>${wildfly-maven-plugin.version}</version>
+    <configuration>
+        <commands>
+            <command>/subsystem=undertow/application-security-domain=other:write-attribute(name=integrated-jaspi, value=false)</command>
+            <command>reload</command>
+        </commands>
+    </configuration>
+</plugin>
+```
+
+The docs for [the WildFly Maven plugin are here](https://docs.wildfly.org/wildfly-maven-plugin/). Except for the cryptic `<command></command>` block, the plugin is pretty simple and easy to use. 
+
+It took a little digging to figure out, but the obscure command block *is* required, at least according to the experts I asked. It disables integrated JASPI in the server and instead delegates validation of credentials to a non-integrated `ServerAuthModule`. This allows for identities to be dynamically created instead of statically stored in an integrated security domain. Take a look at the [Elytron and Java EE Security section of the docs](https://docs.wildfly.org/26/WildFly_Elytron_Security.html#Elytron_and_Java_EE_Security) for more on this.
+
+## Project structure and configuration
+
+The files in the `src` directory are listed below. There are three different services: 
+
+1. the OIDC-protected servlet;
+2. the API servlet protected by a JWT authentication filter; and
+3. an unprotected servlet.
+
+```text
+src
+└── main
+    ├── java
+    │   └── com
+    │       └── demo
+    │           ├── CallbackServlet.java  // OIDC callback handler
+    │           ├── ProtectedServlet.java  // OIDC-handling servlet endpoint
+    │           ├── PublicServlet.java  // Public endpoint
+    │           ├── JwtFilter.java  // Verifies JWT and secures ApiServlet
+    │           ├── ApiServlet.java // API protected by filter
+    │           └── OpenIdConfig.java  // Loads openid.properties
+    ├── resources
+    │   ├── logging.properties  // Simple console logging configuration
+    │   ├── META-INF
+    │   │   ├── beans.xml  // Declare some provided dependencies for deployment
+    │   │   └── MANIFEST.MF  // Configure CDI (Contexts and Dependency Injection)
+    │   └── openid.properties  // OpenID config properties
+    └── webapp
+        └── WEB-INF
+            └── jboss-web.xml  // Configures context root to '/'
```

Comment:
```suggestion
src
└── main
    ├── java
    │   └── com
    │       └── demo
    │           ├── ApiServlet.java // API protected by filter
    │           ├── CallbackServlet.java  // OIDC callback handler
    │           ├── JwtFilter.java  // Verifies JWT and secures ApiServlet
    │           ├── OidcConfig.java  // Loads oidc.properties
    │           ├── ProtectedServlet.java  // OIDC-handling servlet endpoint
    │           └── PublicServlet.java  // Public endpoint
    ├── resources
    │   ├── META-INF
    │   │   ├── beans.xml  // Declare some provided dependencies for deployment
    │   │   └── MANIFEST.MF  // Configure CDI (Contexts and Dependency Injection)
    │   ├── logging.properties  // Simple console logging configuration
    │   └── oidc.properties  // OpenID Connect config properties
    └── webapp
        └── WEB-INF
            └── jboss-web.xml  // Configures context root to '/'
```

Reorder to match IntelliJ and rename so there's no confusion between OpenID and OpenID Connect.